### PR TITLE
[ASA-634][JR] Update application with agency details. Subscribe to agent services.

### PR DESCRIPTION
**Pull request by johnrevill**

### CODE REVIEW

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentOverseasApplicationConnector.scala [hunk 1]
```diff
@@ -21,35 +21,59 @@ import java.net.URL
...
+          case _: NotFoundException => List.empty[OverseasApplication]
+        }
+    }
+  }
+
+  def updateApplicationWithAgencyDetails(
+    agencyDetails: AgencyDetails)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Unit] = {
+    val url = new URL(baseUrl, s"/agent-overseas-application/application").toExternalForm
+
+    monitor(s"ConsumedAPI-agent-overseas-application-PUT") {
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentOverseasApplicationConnector.scala]: update noting acting key word

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.kenshoo.play.metrics.Metrics
+import javax.inject.{Inject, Named, Singleton}
+import play.api.libs.json.{Format, Json}
+import uk.gov.hmrc.agent.kenshoo.monitoring.HttpAPIMonitor
+import uk.gov.hmrc.agentmtdidentifiers.model.Arn
+import uk.gov.hmrc.http._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class OverseasSubscriptionResponse(arn: Arn)
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala]: seems redundant

File: conf/application.conf [hunk 3]
```diff
@@ -74,6 +74,11 @@ microservice {
       port = 9406
     }
 
+    agent-subscription {
```
Comment (arkowalski) [inline conf/application.conf]: will need config for this in app-config-base

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+object OverseasSubscriptionResponse {
+  implicit val formats: Format[OverseasSubscriptionResponse] = Json.format[OverseasSubscriptionResponse]
+}
+
+@Singleton
+class AgentSubscriptionConnector @Inject()(
```
Comment (arkowalski) [inline app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala]: just thinking that this call could be done in agent-overseas-application as we still need to call the complete endpoint to update status.

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala [hunk 5]
```diff
@@ -1,46 +1,77 @@
...
-  def givenSubscriptionApplicationServerError: Unit =
+  def givenApplicationServerError(): Unit =
     stubFor(get(urlEqualTo("/agent-overseas-application/application"))
       .willReturn(aResponse()
         .withStatus(500))
     )
+
+  def givenApplicationUpdateSuccessResponse() =
+    stubFor(put(urlEqualTo("/agent-overseas-application/application")).willReturn(
+      okJson(StubsTestData.acceptedApplication)
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: does this return any json?

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala [hunk 6]
```diff
@@ -1,46 +1,77 @@
...
+        .withStatus(200)
+    ))
+
+  def givenApplicationUpdateNotFoundResponse() =
+    stubFor(put(urlEqualTo("/agent-overseas-application/application")).willReturn(
+      okJson(StubsTestData.acceptedApplication)
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: there should be no body

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala [hunk 7]
```diff
@@ -1,46 +1,77 @@
...
+        .withStatus(404)
+    ))
+
+  def givenApplicationUpdateServerError() =
+    stubFor(put(urlEqualTo("/agent-overseas-application/application")).willReturn(
+      okJson(StubsTestData.acceptedApplication)
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: there should be no body

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentSubscriptionStubs.scala [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        .withStatus(201)
+    ))
+
+  def givenSubscriptionFailed(withStatus: Int): Unit =
+    stubFor(put(urlEqualTo(pathOverseasSubscription))
+      .willReturn(aResponse()
+        .withStatus(withStatus))
+    )
+
+  def givenSubscriptionFailedUnauthorized(): Unit = givenSubscriptionFailed(401)
```
Comment (arkowalski) [inline it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentSubscriptionStubs.scala]: optional -> can remove the () on these for easier use

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala (hunk 2 above)
Comment (johnrevill) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentSubscriptionConnector.scala]: Turns out we need it.

File: app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentOverseasApplicationConnector.scala (hunk 1 above)
Comment (johnrevill) [reply app/uk/gov/hmrc/agentoverseassubscriptionfrontend/connectors/AgentOverseasApplicationConnector.scala]: Updated. I think.

File: conf/application.conf (hunk 3 above)
Comment (johnrevill) [reply conf/application.conf]: Good point, thanks

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala (hunk 5 above)
Comment (johnrevill) [reply it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: Nope

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala (hunk 6 above)
Comment (johnrevill) [reply it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: Good catch

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentSubscriptionStubs.scala (hunk 8 above)
Comment (johnrevill) [reply it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentSubscriptionStubs.scala]: IT's side effecting, so without the () it'd generate compiler warnings

File: it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala (hunk 5 above)
Comment (arkowalski) [reply it/uk/gov/hmrc/agentoverseassubscriptionfrontend/stubs/AgentOverseasApplicationStubs.scala]: in that case remove the body of the response

**Pull request merged**